### PR TITLE
Implement PointsToAnalysis to track locations of all symbols and ins…

### PR DESCRIPTION
**Pull request by mavasani**
…tances and arrays. Primary changes are:

1. Change all the analyses to track values for `AnalysisEntity` instances instead of `ISymbol`. An analysis entity is essentially a symbol (or an allocation or an indexed location) which also has an associated `InstanceLocation` where the entity resides. So for example, each local and parameter symbol with have a single unique `AnalysisEntity` instance, while each member symbol(fields/properties/events/methods) will have separate `AnalysisEntity` instances for every instance within which it is contained.

2. Add support for `PointsToAnalysis`, which tracks the locations that can be pointed to by analysis entities and operations. Note that this points to location is different from the `InstanceLocation` mentioned above, which indicate the location at which an analysis entity _resides_. On the other hand, PointsToAnalysis tracks the location that is _pointed to_ by reference types and allocation operations. Constants and value types always have a value of `PointsToAbstractValue.NoLocation`, indicating that they cannot point to any location.

3. Improve precision of `StringContentsAnalysis` used by DFA rule `CA2100` (Review SQL queries for security vulnerabilities) by using the points to analysis result. This analysis can now track string values of members of named types and arrays. Note that were very minimal changes to `StringContentsAnalysis` in this PR, as the base `DataFlowOperationVisitor` tracks using and manipulating the points to analysis data. This should enable future analyses to consume the points to analysis result to improve precision, without having to write any complex code to handle points to analysis data.

4. Write new unit tests for `CA2100` for precision added by the underlying points to analysis.

### CODE REVIEW

File: src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+                var _ = base.Visit(operation, argument);
+                return PointsToAbstractValue.NoLocation;
+            }
+
+            public override PointsToAbstractValue VisitObjectOrCollectionInitializer(IObjectOrCollectionInitializerOperation operation, object argument)
+            {
+                var _ = base.VisitObjectOrCollectionInitializer(operation, argument);
+
+                // We should have created and created a new points to value for the associated creation operation.
```
Comment (ivanbasov) [inline src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs]: > // We should have created and created a new points to value for the associated creation operation. [](start = 15, length = 99)

please revise this comment and the same below

File: src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+            public override PointsToAbstractValue VisitThrow(IThrowOperation operation, object argument)
+            {
+                var _ = base.VisitThrow(operation, argument);
+                return PointsToAbstractValue.NoLocation;
+            }
+
+            public override PointsToAbstractValue VisitTuple(ITupleOperation operation, object argument)
+            {
+                // TODO: Handle tuples.
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/PointsToAnalysis/PointsToAnalysis.PointsToDataFlowOperationVisitor.cs]: Interesting, I had actually added the work item but forgot to push the commit. Thanks.

---
In reply to: [167042146](https://github.com/dotnet/roslyn-analyzers/pull/1572#discussion_r167042146) [](ancestors = 167042146)

File: src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Framework/DataFlow/AnalysisEntityFactory.cs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+            Debug.Assert(instance != null);
+            Debug.Assert(!indices.IsEmpty);
+            Debug.Assert(elementType != null);
+
+            ISymbol symbol = null;
+            return TryCreate(symbol, indices, elementType, instance, out analysisEntity);
+        }
+
+        private bool TryCreate(ISymbol symbolOpt, ImmutableArray<AbstractIndex> indices,
+            ITypeSymbol type, IOperation instanceOpt, out AnalysisEntity symbolWithLocationinfo)
```
Comment (ivanbasov) [inline src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Framework/DataFlow/AnalysisEntityFactory.cs]: >i [](start = 91, length = 1)

nit: I (capitalized)

**Comment by mavasani** [comment]
I am going to merge this PR so I can work on DisposePointsToAnalysis based on the PointsToAnalysis added in this PR. I will address any feedback on the current PR subsequently.

**Pull request merged**